daemon/cmd/ciliumendpoints.go

```diff
@@ -0,0 +1,116 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/cilium/cilium/pkg/endpoint"
+	cilium_v2a1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+)
+
+type endpointCache interface {
+	LookupPodName(name string) *endpoint.Endpoint
+}
+
+// Used by cleanStaleCEPs to retrieve ceps/pods to do cleanup of stale ceps.
+// Values retrieved only used as keys for accessing other caches and performing
+// deletes.
+func (d *Daemon) getStore(name string) cache.Store {
+	return d.k8sWatcher.GetStore(name)
+}
+
+// cleanStaleCEPs runs on Daemon init, attempting to delete any CiliumEndpoints that are referencing local
+// Pods that are not being managed (i.e. do not have Endpoints).
+// This must only be run after K8s Pod and CES/CEP caches are synced and local endpoint restoration is complete.
+func (d *Daemon) cleanStaleCEPs(ctx context.Context, eps endpointCache, ciliumClient ciliumv2.CiliumV2Interface, disableCEPCRD bool) error {
+	keyFn := func(ns string, name string) string { return ns + "/" + name }
+
+	cesContainedCEPLookup := map[string]*cilium_v2a1.CoreCiliumEndpoint{}
+	if disableCEPCRD {
+		cesObjs := d.getStore("ciliumendpointslice").List()
+		for _, cesObj := range cesObjs {
+			ces, ok := cesObj.(*cilium_v2a1.CiliumEndpointSlice)
+			if !ok {
+				return fmt.Errorf("unexpected object type returned from ciliumendpointslice store")
+			}
+			for i := range ces.Endpoints {
+				cep := ces.Endpoints[i]
+				if cep.Networking.NodeIP == node.GetCiliumEndpointNodeIP() {
+					cesContainedCEPLookup[keyFn(ces.Namespace, cep.Name)] = &cep
+				}
+			}
+		}
+	}
+
+	// GetCachedPods will only return local pods in cases where ciliumendpoint CRD is disabled.
+	pods, err := d.k8sWatcher.GetCachedPods()
+	if err != nil {
+		return fmt.Errorf("could not get pods from local cache: %w", err)
+	}
+
+	for _, pod := range pods {
+		if pod.Spec.HostNetwork {
+			continue
+		}
+
+		var cepName, cepNamespace string
+		store := d.getStore("ciliumendpoint")
+		if store == nil {
+			cep, exists := cesContainedCEPLookup[keyFn(pod.Namespace, pod.Name)]
+			if !exists {
+				continue
+			}
+			cepName = cep.Name
+			cepNamespace = pod.Namespace
+		} else {
+			cepObj, exists, err := d.getStore("ciliumendpoint").GetByKey(keyFn(pod.Namespace, pod.Name))
```

Comment:
could this read `cepObj, exists, err := store.GetByKey(keyFn(pod.Namespace, pod.Name))` instead?

Reply:
could also set up the key before the if blocks as it is the same for both cases?